File: #1028.recover-a-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* recoverFromPreorder(string traversal) {

        int size = traversal.size();
        int level;
        int value;
        TreeNode* node;
        stack<TreeNode*> stk;
        for (int i = 0; i < size;) {
            level = 0;
            value = 0;

            for (; traversal[i]=='-' && i < size; i++) {
                level++;
            }

            for (; traversal[i]!='-' && i < size; i++) {
                value = value * 10 + (traversal[i] - '0');
            }

            node = new TreeNode(value);
            while (stk.size() > level) {
                stk.pop();
            }

            if (!stk.empty()) {
                if (!stk.top()->left) {
                    stk.top()->left = node;
                } else {
                    stk.top()->right = node;
                }
            }

            stk.push(node);
        }
        
        while (!stk.empty()) {
            node = stk.top();
            stk.pop();
        }

        return node;
    }
};

File: #1028.recover-a-tree-from-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {string} traversal
 * @return {TreeNode}
 */
var recoverFromPreorder = function(traversal) {
    const map = {};
    const reg = /((-+|^)(\d+))/g;
    let depth = 0;
    let matches;
    while (matches = reg.exec(traversal)) {
        depth = matches[2].length;
        const curr = new TreeNode(parseInt(matches[3]));
        const parent = map[depth -1];
        if (parent) {
            if (parent.left === null) {
                parent.left = curr;
            } else {
                parent.right = curr;
            }
        }
        map[depth] = curr;
    }
    return map[0];
};
